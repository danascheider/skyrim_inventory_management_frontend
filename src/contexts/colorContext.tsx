import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
