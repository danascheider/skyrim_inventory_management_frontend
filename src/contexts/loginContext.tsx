import { createContext, useState, useEffect, useCallback } from 'react'
import { type User } from 'firebase/auth'
import { useAuthState } from 'react-firebase-hooks/auth'
import { auth, signOutWithGoogle } from '../firebase'
import { ProviderProps } from '../types/contexts'

export interface LoginContextType {
  authLoading: boolean
  token: string | null
  withTokenRefresh: (fn: (idToken: string) => void) => void
  user?: User | null
}

export const LoginContext = createContext<LoginContextType>({
  user: null,
  token: null,
  authLoading: true,
  withTokenRefresh: () => {},
})

export const LoginProvider = ({ children }: ProviderProps) => {
  const [user, authLoading, authError] = useAuthState(auth)
  const [token, setToken] = useState<string | null>(null)

  const refreshToken = useCallback(() => {
    if (user) user.getIdToken(true).then((idToken) => setToken(idToken))
  }, [user])

  const withTokenRefresh = (fn: Function) => {
    if (user) {
      user.getIdToken(true).then((idToken) => {
        setToken(idToken)
        fn(idToken)
      })
    }
  }

  const value = {
    user,
    token,
    withTokenRefresh,
    authLoading,
  }

  useEffect(() => {
    if (authError) signOutWithGoogle()

    refreshToken()
  }, [user, authError])

  return <LoginContext.Provider value={value}>{children}</LoginContext.Provider>
}
