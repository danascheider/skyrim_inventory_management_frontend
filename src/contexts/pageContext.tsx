import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
