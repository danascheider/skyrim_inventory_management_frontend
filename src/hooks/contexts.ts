import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )