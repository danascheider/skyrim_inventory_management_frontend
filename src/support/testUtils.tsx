import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
