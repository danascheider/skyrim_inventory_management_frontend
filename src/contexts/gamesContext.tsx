import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
