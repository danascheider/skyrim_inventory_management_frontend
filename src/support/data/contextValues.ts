import { type GamesContextType } from '../../contexts/gamesContext'
import { type LoginContextType } from '../../contexts/loginContext'
import { type ShoppingListsContextType } from '../../contexts/shoppingListsContext'
import { DONE, ERROR, LOADING } from '../../utils/loadingStates'
import { testUser } from './users'
import { allGames } from './games'
import { emptyShoppingLists, shoppingListsForGame } from './shoppingLists'

const noop = () => {}

/**
 *
 * Default values for Login context
 *
 */

export const loginContextValue: LoginContextType = {
  user: testUser,
  token: 'xxxxxxx',
  requireLogin: noop,
  authLoading: false,
}

export const loadingLoginContextValue: LoginContextType = {
  user: null,
  token: null,
  requireLogin: noop,
  authLoading: true,
}

export const unauthenticatedLoginContextValue: LoginContextType = {
  user: null,
  token: null,
  requireLogin: noop,
  authLoading: false,
}

/**
 *
 * Default value for Games context
 *
 */

export const gamesContextValue: GamesContextType = {
  games: allGames,
  gamesLoadingState: DONE,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

export const gamesContextValueLoading: GamesContextType = {
  games: [],
  gamesLoadingState: LOADING,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

export const gamesContextValueError: GamesContextType = {
  games: [],
  gamesLoadingState: ERROR,
  createGame: noop,
  updateGame: noop,
  destroyGame: noop,
}

/**
 *
 * Default values for Shopping Lists context
 *
 */

export const shoppingListsContextValueEmpty: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: DONE,
  createShoppingList: noop,
}

export const shoppingListsContextValue: ShoppingListsContextType = {
  shoppingLists: shoppingListsForGame(77),
  shoppingListsLoadingState: DONE,
  createShoppingList: noop,
}

export const shoppingListsContextValueLoading: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: LOADING,
  createShoppingList: noop,
}

export const shoppingListsContextValueError: ShoppingListsContextType = {
  shoppingLists: emptyShoppingLists,
  shoppingListsLoadingState: ERROR,
  createShoppingList: noop,
}
