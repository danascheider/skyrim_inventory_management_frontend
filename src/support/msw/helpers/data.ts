import {
  type RequestShoppingListItem,
  type RequestShoppingList,
  type ResponseShoppingListItem,
  type ResponseShoppingList,
} from '../../../types/apiData'
import { allGames } from '../../data/games'
import {
  allShoppingLists,
  shoppingListsForGame,
} from '../../data/shoppingLists'

const gameIds = allGames.map(({ id }) => id)

/**
 *
 * Shopping List Creation
 *
 */

export const newShoppingList = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (!existingLists.length)
    throw new Error(
      'Cannot generate single list for game without existing aggregate'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: existingLists[0].id,
      title: attributes.title || 'New Shopping List',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

export const newShoppingListWithAggregate = (
  attributes: RequestShoppingList,
  gameId: number
): ResponseShoppingList[] => {
  if (gameIds.indexOf(gameId) < 0)
    throw new Error(
      'Cannot generate shopping list for game that does not exist in test data'
    )

  const existingLists = shoppingListsForGame(gameId)

  if (existingLists.length)
    throw new Error(
      'Cannot generate new aggregate list for game that already has one'
    )

  return [
    {
      id: 93,
      game_id: gameId,
      aggregate: true,
      aggregate_list_id: null,
      title: 'All Items',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
    {
      id: 94,
      game_id: gameId,
      aggregate: false,
      aggregate_list_id: 93,
      title: attributes.title || 'My Shopping List 1',
      list_items: [],
      created_at: new Date(),
      updated_at: new Date(),
    },
  ]
}

/**
 *
 * Shopping list item creation
 *
 */

export const newShoppingListItem = (
  attributes: RequestShoppingListItem,
  listId: number
) => {
  const list = allShoppingLists.find(({ id }) => id === listId)

  if (!list)
    throw new Error(`No shopping list with ID ${listId} in the test data`)

  const shoppingList = { ...list }
  const allLists = shoppingListsForGame(shoppingList.game_id)
  const aggregateList = { ...allLists[0] }

  const newItem: ResponseShoppingListItem = {
    id: 42,
    list_id: listId,
    description: 'Dummy description for TypeScript',
    quantity: 1,
    unit_weight: null,
    notes: null,
    created_at: new Date(),
    updated_at: new Date(),
    ...attributes,
  }

  const newAggregateListItem: ResponseShoppingListItem = {
    ...newItem,
    id: 43,
    list_id: aggregateList.id,
  }

  aggregateList.list_items = [newAggregateListItem, ...aggregateList.list_items]
  shoppingList.list_items = [newItem, ...shoppingList.list_items]

  return [aggregateList, shoppingList]
}
