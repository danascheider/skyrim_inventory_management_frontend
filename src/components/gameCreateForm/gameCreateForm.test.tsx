import { describe, test, expect, vitest } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { RequestGame } from '../../types/apiData'
import { renderAuthenticated } from '../../support/testUtils'
import { gamesContextValue } from '../../support/data/contextValues'
import { PageProvider } from '../../contexts/pageContext'
import { GamesContext } from '../../contexts/gamesContext'
import GameCreateForm from './gameCreateForm'

describe('<GameCreateForm />', () => {
  describe('when enabled', () => {
    test('displays the correct fields', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <GameCreateForm />
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper).toBeTruthy()

      expect(wrapper.getByText('Create Game...')).toBeTruthy()
      expect(wrapper.getByTestId('gameCreateFormForm')).toBeTruthy()
    })

    test('matches snapshot', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <GameCreateForm />
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper).toMatchSnapshot()
    })

    describe('creating a game', () => {
      describe('when the form has no values', () => {
        test('calls the createGame function on the Games provider', () => {
          const createGame = vitest
            .fn()
            .mockImplementation((_game: RequestGame) => {})

          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesContext.Provider
                value={{ ...gamesContextValue, createGame }}
              >
                <GameCreateForm />
              </GamesContext.Provider>
            </PageProvider>
          )

          const form = wrapper.getByTestId('gameCreateFormForm')

          act(() => {
            fireEvent.submit(form)
          })

          expect(createGame).toHaveBeenCalledWith(
            { name: null, description: null },
            expect.any(Function)
          )
        })
      })

      describe('when the form has been filled out', () => {
        test('calls the createGame function on the Games provider', () => {
          const createGame = vitest
            .fn()
            .mockImplementation((_game: RequestGame) => {})
          const destroyGame = () => {}
          const updateGame = () => {}

          const wrapper = renderAuthenticated(
            <PageProvider>
              <GamesContext.Provider
                value={{ ...gamesContextValue, createGame }}
              >
                <GameCreateForm />
              </GamesContext.Provider>
            </PageProvider>
          )

          const nameInput = wrapper.getByTestId('createNameField')
          const descInput = wrapper.getByTestId('createDescriptionField')
          const form = wrapper.getByTestId('gameCreateFormForm')

          act(() => {
            fireEvent.change(nameInput, { target: { value: 'Skyrim' } })
            fireEvent.change(descInput, {
              target: { value: 'Custom description' },
            })
            fireEvent.submit(form)
          })

          expect(createGame).toHaveBeenCalledWith(
            { name: 'Skyrim', description: 'Custom description' },
            expect.any(Function)
          )
        })
      })
    })
  })

  // Again, because of limitations on DOM testing with Vitest, we are limited to
  // snapshot testing for this component state. There is a promising package,
  // https://github.com/chaance/vitest-dom, that forks testing-library/js-dom and
  // introduces its matches for Vitest. However, this package is only available on
  // version 0.0.4 and has a peer dependency for Vitest "^0.16.0", when we are
  // on 0.29.2, so it's too new a package to really rely on.
  describe('when disabled', () => {
    test('matches snapshot', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <GameCreateForm disabled />
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
