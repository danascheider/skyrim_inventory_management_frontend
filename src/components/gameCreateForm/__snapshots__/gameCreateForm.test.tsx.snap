// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<GameCreateForm /> > when disabled > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_83bedb"
        style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
      >
        <h3
          aria-controls="gameCreateForm"
          aria-expanded="false"
          class="_toggle_83bedb"
          role="button"
        >
          Create Game...
        </h3>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="gameCreateForm"
          style="height: 0px; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <form
              class="_form_83bedb"
              data-testid="gameCreateFormForm"
            >
              <fieldset
                class="_fieldset_83bedb"
              >
                <input
                  aria-label="Name"
                  class="_input_83bedb"
                  data-testid="nameField"
                  disabled=""
                  name="name"
                  pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                  placeholder="Name"
                  title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                  type="text"
                />
              </fieldset>
              <fieldset
                class="_fieldset_83bedb"
              >
                <input
                  aria-label="Description"
                  class="_input_83bedb"
                  data-testid="descriptionField"
                  disabled=""
                  name="description"
                  placeholder="Description"
                  type="text"
                />
              </fieldset>
              <button
                class="_button_83bedb"
                data-testid="createGameSubmit"
                disabled=""
                type="submit"
              >
                Create
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_83bedb"
      style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
    >
      <h3
        aria-controls="gameCreateForm"
        aria-expanded="false"
        class="_toggle_83bedb"
        role="button"
      >
        Create Game...
      </h3>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="gameCreateForm"
        style="height: 0px; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <form
            class="_form_83bedb"
            data-testid="gameCreateFormForm"
          >
            <fieldset
              class="_fieldset_83bedb"
            >
              <input
                aria-label="Name"
                class="_input_83bedb"
                data-testid="nameField"
                disabled=""
                name="name"
                pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                placeholder="Name"
                title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                type="text"
              />
            </fieldset>
            <fieldset
              class="_fieldset_83bedb"
            >
              <input
                aria-label="Description"
                class="_input_83bedb"
                data-testid="descriptionField"
                disabled=""
                name="description"
                placeholder="Description"
                type="text"
              />
            </fieldset>
            <button
              class="_button_83bedb"
              data-testid="createGameSubmit"
              disabled=""
              type="submit"
            >
              Create
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<GameCreateForm /> > when enabled > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_83bedb"
        style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
      >
        <h3
          aria-controls="gameCreateForm"
          aria-expanded="false"
          class="_toggle_83bedb"
          role="button"
        >
          Create Game...
        </h3>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="gameCreateForm"
          style="height: 0px; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <form
              class="_form_83bedb"
              data-testid="gameCreateFormForm"
            >
              <fieldset
                class="_fieldset_83bedb"
              >
                <input
                  aria-label="Name"
                  class="_input_83bedb"
                  data-testid="nameField"
                  name="name"
                  pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                  placeholder="Name"
                  title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                  type="text"
                />
              </fieldset>
              <fieldset
                class="_fieldset_83bedb"
              >
                <input
                  aria-label="Description"
                  class="_input_83bedb"
                  data-testid="descriptionField"
                  name="description"
                  placeholder="Description"
                  type="text"
                />
              </fieldset>
              <button
                class="_button_83bedb"
                data-testid="createGameSubmit"
                type="submit"
              >
                Create
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_83bedb"
      style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
    >
      <h3
        aria-controls="gameCreateForm"
        aria-expanded="false"
        class="_toggle_83bedb"
        role="button"
      >
        Create Game...
      </h3>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="gameCreateForm"
        style="height: 0px; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <form
            class="_form_83bedb"
            data-testid="gameCreateFormForm"
          >
            <fieldset
              class="_fieldset_83bedb"
            >
              <input
                aria-label="Name"
                class="_input_83bedb"
                data-testid="nameField"
                name="name"
                pattern="^\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*$"
                placeholder="Name"
                title="Name can contain only alphanumeric characters, spaces, commas, hyphens, and apostrophes"
                type="text"
              />
            </fieldset>
            <fieldset
              class="_fieldset_83bedb"
            >
              <input
                aria-label="Description"
                class="_input_83bedb"
                data-testid="descriptionField"
                name="description"
                placeholder="Description"
                type="text"
              />
            </fieldset>
            <button
              class="_button_83bedb"
              data-testid="createGameSubmit"
              type="submit"
            >
              Create
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
